File: server-express/src/utils/jwtUtil.ts
import jwt from "jsonwebtoken";
import { Request, Response, NextFunction } from "express";
import { AuthenticatedRequest } from "../types/express";
import { InvalidRefreshToknError, InvalidTokenError, NoRefreshTokenError, NoTokenError } from "../Exceptions/jwt-error";

const ACCESS_TOKEN_EXPIRY = "2h";
const REFRESH_TOKEN_EXPIRY = "7d";

interface TokenPayload {
    id: string;
    role: string;
}

/**
 * Generates an access token
 */
export const generateAccessToken = (payload: TokenPayload): string => {
    return jwt.sign(payload, process.env.JWT_SECRET as string, {
        expiresIn: ACCESS_TOKEN_EXPIRY,
    });
};

/**
 * Generates a refresh token
 */
export const generateRefreshToken = (payload: TokenPayload): string => {
    return jwt.sign(payload, process.env.REFRESH_TOKEN_SECRET as string, {
        expiresIn: REFRESH_TOKEN_EXPIRY,
    });
};

/**
 * Verifies the access token and attaches the user to the request
 */
export const verifyAccessToken = (req: AuthenticatedRequest): TokenPayload => {
    const token = req.headers.authorization?.split(' ')[1];

    if (!token) {
        throw new NoTokenError('Access token is required');
    }

    try {
        return jwt.verify(token, process.env.JWT_SECRET as string) as TokenPayload;
    } catch (error) {
        throw new InvalidTokenError('Invalid or expired token');
    }
};

/**
 * Verifies the refresh token from cookies
 */
export const verifyRefreshToken = (req: AuthenticatedRequest): TokenPayload => {
    const refreshToken = req.cookies?.refreshToken;

    if (!refreshToken) {
        throw new NoRefreshTokenError('Refresh token is required');
    }

    try {
        return jwt.verify(refreshToken, process.env.REFRESH_TOKEN_SECRET as string) as TokenPayload;
    } catch (error) {
        throw new InvalidRefreshToknError('Invalid or expired token');
    }
};

/**
 * Extracts the user role from a token
 */
export const extractUserRole = (token: string): string => {
    try {
        const decoded = jwt.decode(token) as TokenPayload | null;
        if (!decoded || !decoded.role) {
            throw new Error("Invalid token or role not found");
        }
        return decoded.role;
    } catch (error) {
        throw new InvalidTokenError("Failed to extract user role");
    }
};

File: server-springboot/netflix-backend-springboot/src/main/java/com/example/netflix_backend_springboot/util/JwtUtil.java
package com.example.netflix_backend_springboot.util;

import io.jsonwebtoken.*;
import org.springframework.stereotype.Component;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Component
public class JwtUtil {

    private final String ACCESS_TOKEN_SECRET = "your_access_token_secret";
    private final String REFRESH_TOKEN_SECRET = "your_refresh_token_secret";
    private final long ACCESS_TOKEN_EXPIRY = 1000 * 60 * 60 * 2; // 2 hours
    private final long REFRESH_TOKEN_EXPIRY = 1000 * 60 * 60 * 24 * 7; // 7 days

    // Generates an access token with the given payload and expiry time
    public String generateAccessToken(Map<String, Object> payload) {
        return Jwts.builder()
                .setClaims(payload)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + ACCESS_TOKEN_EXPIRY))
                .signWith(SignatureAlgorithm.HS256, ACCESS_TOKEN_SECRET)
                .compact();
    }

    // Generates a refresh token with the given payload and expiry time
    public String generateRefreshToken(Map<String, Object> payload) {
        return Jwts.builder()
                .setClaims(payload)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + REFRESH_TOKEN_EXPIRY))
                .signWith(SignatureAlgorithm.HS256, REFRESH_TOKEN_SECRET)
                .compact();
    }

    // Generates an access token with username and role claims
    public String generateTokenWithClaims(String username, String role) {
        Map<String, Object> claims = new HashMap<>();
        claims.put("role", role);
        claims.put("sub", username);
        return generateAccessToken(claims);
    }

    // Verifies the validity of the access token and returns its claims
    public Claims verifyAccessToken(String token) {
        try {
            return Jwts.parser()
                    .setSigningKey(ACCESS_TOKEN_SECRET)
                    .parseClaimsJws(token)
                    .getBody();
        } catch (JwtException e) {
            throw new RuntimeException("Invalid or expired access token");
        }
    }

    // Verifies the validity of the refresh token and returns its claims
    public Claims verifyRefreshToken(String token) {
        try {
            return Jwts.parser()
                    .setSigningKey(REFRESH_TOKEN_SECRET)
                    .parseClaimsJws(token)
                    .getBody();
        } catch (JwtException e) {
            throw new RuntimeException("Invalid or expired refresh token");
        }
    }

    // Extracts claims from a refresh token
    public Map<String, Object> extractClaimsFromRefreshToken(String token) {
        Claims claims = verifyRefreshToken(token);
        Map<String, Object> extractedClaims = new HashMap<>();
        extractedClaims.put("username", claims.getSubject());
        extractedClaims.put("role", claims.get("role", String.class));
        return extractedClaims;
    }

    // Extracts the user role from the access token
    public String extractUserRole(String token) {
        try {
            Claims claims = Jwts.parser()
                    .setSigningKey(ACCESS_TOKEN_SECRET)
                    .parseClaimsJws(token)
                    .getBody();
            return claims.get("role", String.class);
        } catch (Exception e) {
            throw new RuntimeException("Failed to extract user role");
        }
    }

    // Extracts the username from the JWT token
    public String extractUsername(String token) {
        try {
            Claims claims = Jwts.parser()
                    .setSigningKey(ACCESS_TOKEN_SECRET)
                    .parseClaimsJws(token)
                    .getBody();
            return claims.getSubject();
        } catch (Exception e) {
            throw new RuntimeException("Failed to extract username from token");
        }
    }
}
